core/src/main/java/org/elasticsearch/action/bulk/Retry.java

```diff
@@ -55,42 +66,48 @@ public Retry policy(BackoffPolicy backoffPolicy) {
         return this;
     }
 
-    Retry(Class<? extends Throwable> retryOnThrowable) {
-        this.retryOnThrowable = retryOnThrowable;
+    /**
+     * @param threadPool The threadPool that will be used to schedule retries.
+     */
+    public Retry using(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+        return this;
     }
 
     /**
-     * Invokes #bulk(BulkRequest, ActionListener) on the provided client. Backs off on the provided exception and delegates results to the
-     * provided listener.
-     *
-     * @param client      Client invoking the bulk request.
+     * Invokes #apply(BulkRequest, ActionListener). Backs off on the provided exception and delegates results to the
+     * provided listener. Retries will be attempted using the provided schedule function
+     * @param consumer The consumer to which apply the request and listener
      * @param bulkRequest The bulk request that should be executed.
-     * @param listener    A listener that is invoked when the bulk request finishes or completes with an exception. The listener is not
+     * @param listener A listener that is invoked when the bulk request finishes or completes with an exception. The listener is not
+     * @param settings settings
      */
-    public void withAsyncBackoff(Client client, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
-        AsyncRetryHandler r = new AsyncRetryHandler(retryOnThrowable, backoffPolicy, client, listener);
+    public void withAsyncBackoff(BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer, BulkRequest bulkRequest, ActionListener<BulkResponse> listener, Settings settings) {
```

Comment:
I dunno. I don't love this class. I was trying to keep it closer as it used to be, and only replace Client with biconsumer and threadpool. I also don't think it's clear from the method name that `withSyncBackoff` actually sends the request. But I don't want to change too much as part of this PR either. If you feel like it's better as-is I am fine with it too.